split MagnetSet and MagnetCoil

Edgar-21 opened the pull request:
Round 2 at splitting this class to add the custom geometry functionality while preserving all the other features. After being away from this for a bit I realized I wanted to do it differently than in #173 and it was easier just to start fresh.

MagnetSet() is the parent class, and MagnetSetFromFilaments and MagnetSetFromGeometry inherit from it. MagnetSetFromFilaments is functionally the same as the current MagnetSet class. MagnetSetFromGeometry enables use of existing CAD files with the typical parastell workflows. Both classes have tested to work with DAGMC model generation and with the radial distance utilities.

MagnetSetFromGeometry has no need for MagnetCoil objects, so a new object, Filament, has been introduced which contains the functionality needed for radial distance finding with both MagnetSetFromFilaments and MagnetSetFromGeometry.

Kind of a big diff but to get the tests passing many little things needed to be tweaked. There is very little new code here, mostly just moving around existing pieces.

Addresses #174

connoramoreno commented on parastell/magnet_coils.py:
```diff
@@ -169,33 +284,6 @@ def _compute_radial_distance_data(self):
-        # Sort coils by center-of-mass toroidal angle and overwrite stored list
-        self.magnet_coils = self.sort_coils_toroidally()
-
     def _cut_magnets(self):
         """Cuts the magnets at the planes defining the toriodal extent.
         (Internal function not intended to be called externally)
```
Would this not also be useful for imported geometries?

Edgar-21 replied:
I'm assuming that people who bring their own cad will have it in state where it matches their geometry, or will have the means to do that.

connoramoreno reviewed (changes requested):
Here are some initial thoughts - we can discuss these further if you disagree. This branch will also need to be rebased onto `main`.

Edgar-21 commented on parastell/magnet_coils.py:
```diff
@@ -238,141 +352,183 @@ def build_magnet_coils(self):
 
-        self.volume_ids = list(range(first_vol_id, last_vol_id + 1))
+    Arguments:
+        coils_file (str): path to coil filament data file.
```
I agree, I'll tackle updating rdu to work with both subclasses in an upcoming PR, I've created an issue and assigned myself here #186

Edgar-21 commented on parastell/magnet_coils.py:
```diff
@@ -238,141 +352,183 @@ def build_magnet_coils(self):
-    def export_step(self, step_filename="magnet_set", export_dir=""):
-        """Export CAD solids as a STEP file via CadQuery.
+    Optional attributes:
+        start_line (int): starting line index for data in filament data file
+            (defaults to 3).
+        sample_mod (int): sampling modifier for filament points (defaults to
+            1). For a user-defined value n, every nth point will be sampled.
+        scale (float): a scaling factor between the units of the point-locus
+            data and [cm] (defaults to m2cm = 100).
```
yes, with removing the filament data from this subclass that makes sense

Edgar-21 commented on parastell/magnet_coils.py:
```diff
@@ -238,141 +352,183 @@ def build_magnet_coils(self):
-        self.volume_ids = list(range(first_vol_id, last_vol_id + 1))
+    Arguments:
+        coils_file (str): path to coil filament data file.
+        geometry_file (str): filename of the existing coil geometry. Can be of
+            the types supported by cubit_io.import_geom_to_cubit()
+        working_dir (str): path to directory in which existing geometry
+            is saved.
```
yeah seems like a good plan

connoramoreno reviewed (changes requested):
Just a quick thing on Cubit syntax for newer versions. We can (potentially) submit an additional review after Friday's meeting.

Edgar-21 commented on parastell/magnet_coils.py:
```diff
@@ -16,6 +16,99 @@
+            filename=mesh_filename, export_dir=export_dir
+        )
+
+    def sort_filaments_toroidally(self):
```
move this this to MagnetSetFromFilaments()

connoramoreno commented:
@gonuke, Edgar and I discussed the software design of the code changes in `magnet_coils.py` and we'd like your input on best practices/pythonicity. I'm mostly on board with the changes Edgar is introducing here but I'm hesitant to approve this PR with the current implementation of the parent class `MagnetSet`. Specifically, `MagnetSet` is not functional as a stand-alone class. Although it's a parent class, it relies on member variables, including `self.geometry_file` and `self.working_dir` in the `import_geom_cubit` method, being set by one of the child classes. Is that okay? I'm unsure what "proper" protocol is here.

gonuke commented:
> @gonuke, Edgar and I discussed the software design of the code changes in `magnet_coils.py` and we'd like your input on best practices/pythonicity. I'm mostly on board with the changes Edgar is introducing here but I'm hesitant to approve this PR with the current implementation of the parent class, `MagnetSet`. Specifically, `MagnetSet` is not functional as a stand-alone class. Although it's a parent class, it relies on member variables, including `self.geometry_file` and `self.working_dir` in the `import_geom_cubit` method, being set by one of the child classes. Is that okay? I'm unsure what "proper" code design is here.

In general, this is a common pattern.  For example, in C++, it's common to have a parent class that defines a common API (i.e. a set of functions) but define those function in the parent class to be explicitly undefined so that it's never legal to define an object from that parent class.

I'll take a closer look at this specific implementation to see if I have any thoughts, though.

Edgar-21 commented on parastell/cubit_io.py:
```diff
@@ -155,9 +172,27 @@ def export_dagmc_cubit(
 
     export_path = Path(export_dir) / Path(filename).with_suffix(".h5m")
-    cubit.cmd(f'export dagmc "{export_path}" overwrite')
+    cubit.cmd(f'export cf_dagmc "{export_path}" overwrite')
```
I suppose I can upgrade, this has now been reverted

connoramoreno commented on tests/test_magnet_coils.py:
```diff
@@ -2,42 +2,95 @@
 import parastell.magnet_coils as magnet_coils
 from parastell.cubit_io import create_new_cubit_instance
 
+files_to_remove = [
+    "magnet_set.step",
+    "magnet_mesh.exo",
+    "magnet_mesh.h5m",
+    "stellarator.log",
+    "step_import.log",
+    "single_coil.step",
+]
+
 
 def remove_files():
+    for file in files_to_remove:
+        if Path(file).exists():
+            Path.unlink(file)
+
```
Nice

connoramoreno replied:
I'd like to change all of the tests to use this routine. I think a separate PR would be most appropriate for that though.

connoramoreno reviewed (commented):
This is looking pretty good to me now, thanks @Edgar-21. I'll hold off on approving/merging until Paul gives his thoughts on our software design discussion

gonuke commented on .gitignore:
```diff
@@ -17,6 +17,7 @@
 *.yaml
 !config.yaml
 *.lic
+!magnet_geom.step
```
And maybe a comment for any of the exceptinos...

Edgar-21 replied:
grouped the test files together

gonuke reviewed (changes requested):
I think the code looks OK.  It's a little hard to track all the changes because many of them are just a reorganization of existing code that are hard to compare because of how they have moved in the file.  It seems OK to me.

I have one minor request on the organization of the

Edgar-21 commented:
rebased, and made a few changes to get things working with the new cad to dagmc methods. I've added a method to InvesselBuild() and MagnetSet() that adds the geometry to a cad_to_dagmc instance, rather than getting the cad solids from each class and combining them in Stellarator.build_cad_to_dagmc_model()

connoramoreno commented on parastell/parastell.py:
```diff
@@ -441,29 +441,13 @@ def build_cad_to_dagmc_model(self):
 
         if self.invessel_build:
-            ivb_solids, ivb_material_names = (
-                self.invessel_build.extract_solids_and_mat_tags()
-            )
-            solids.extend(ivb_solids)
-            material_names.extend(ivb_material_names)
+            self.invessel_build.add_solids_to_cad_to_dagmc(self.dagmc_model)
 
         if self.magnet_set:
-            ms_solids, ms_material_names = (
-                self.magnet_set.extract_solids_and_mat_tag()
-            )
-            solids.extend(ms_solids)
-            material_names.extend(ms_material_names)
-
-        self.dagmc_model = cad_to_dagmc.CadToDagmc()
-
-        for solid, mat_tag in zip(solids, material_names):
-            self.dagmc_model.add_cadquery_object(
-                solid, material_tags=[mat_tag]
-            )
+            self.magnet_set.add_solids_to_cad_to_dagmc(self.dagmc_model)
```
Let's modify this structure to make it more transparent what's happening in these methods (per our discussion in Hacker Within).

Edgar-21 replied:
donezo

connoramoreno reviewed (changes requested):
A comment about code structure, almost done!

connoramoreno reviewed (approved):
Looks great! Thanks for this upgrade @Edgar-21

connoramoreno commented:
Just need @gonuke to approve before we can merge.

gonuke reviewed (approved):
Thanks @Edgar-21
